### PR TITLE
Midi/track input

### PR DESCRIPTION
**Pull request by qm210**
I just rebased your current upstream master and I _think_ I made it work again, but...

Expecting your feedback ;)

### CODE REVIEW

File: tracker/model.go [hunk 1]
```diff
@@ -496,36 +509,6 @@ func (m *Model) fixIDCollisions() {
 	}
 }
 
-var validParameters = map[string](map[string]bool){}
```
Comment (vsariola) [inline tracker/model.go:499]: You rebased to master by essentially just ignoring all the changes in master :)

File: tracker/processor.go [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+	"github.com/vsariola/sointu"
+)
+
+type Processor struct {
+	*Player
+	playerProcessContext PlayerProcessContext
+	uiProcessor          EventProcessor
+}
+
+func CreateProcessor(player *Player, context PlayerProcessContext, uiProcessor EventProcessor) *Processor {
```
Comment (vsariola) [inline tracker/processor.go:13]: Anything that looks like a constructor (taking bunch of parameters and returning a pointer to an object allocated in the heap) is usually called NewXXXX; here I would call this method NewProcessor

File: tracker/table.go [hunk 3]
```diff
@@ -149,8 +156,8 @@ func (v Table) SetCursorY(y int) {
 
 // Order methods
 
-func (v *Order) Table() Table {
```
Comment (vsariola) [inline tracker/table.go:152]: This seems unrelated to the main commit, please use `style:` commits if you think there's really need to harmonize receiver names.

File: tracker/table.go [hunk 4]
```diff
@@ -397,6 +400,41 @@ func (e *Order) Title(x int) (title string) {
...
+		return trackIndex, trackIndex, err1
+	}
+	lastIndex, err2 := m.d.Song.Patch.InstrumentForVoice(lastVoice)
+	if err2 != nil {
+		return trackIndex, trackIndex, err2
+	}
+	return firstIndex, lastIndex, nil
+}
+
+func (m *Order) TrackIndicesForCurrentInstrument() []int {
```
Comment (vsariola) [inline tracker/table.go:418]: Please use iterators instead of returning slices (which should become easy once AllTracksWithSameInstrument has an iterator signature as well)

File: tracker/table.go [hunk 5]
```diff
@@ -397,6 +400,41 @@ func (e *Order) Title(x int) (title string) {
...
+	count := 0
+	for _, t := range m.TrackIndicesForCurrentInstrument() {
+		if t > currentTrack {
+			count++
+		}
+	}
+	return count
+}
+
+func (m *Order) CurrentTrack() int {
```
Comment (vsariola) [inline tracker/table.go:434]: Is this really needed? `order.CurrentTrack()` vs. `order.Cursor().X` ?

**Review by vsariola — COMMENTED** [review]
Cool stuff, accepting notes through the MIDI while jamming seems to be the point of this?  There was so much changes to many different files, so it's a bit difficult to review what it actually does.

1. Please properly rebase to the master. You have essentially just reverted all the changes there :) In particular, the refactorings e.g. MIDIContexter -> MIDIContext, m.d.Loop -> m.Loop and you removed again the fixUnitParams...
2. There's a ton of minor style changes (e.g. renaming method receivers), documentation (comments), and changing .gitignore. These probably should go to separate commits, with `style:`, `doc:` and `chore:` type indicators.
3. I'd appreciate if you can already squash the commits, because it seems that there's more different types of changes than one `feat:` commit here. I recommend you making a backup branch of your current commit history, if you start removing some of those unrelated style/doc/chore changes from the PR.
4. Please add a message in the CHANGELOG.md (on the first line after "Added") to describe what feature you actually added.

**Comment by vsariola** [comment]
Oh, and all the VST binaries don't seem to build, that would have to be sorted out also.

**Comment by vsariola** [comment]
So, I got to experiment with this properly. I have a question: normally, when recording and jamming, we interpret the midi channel to correspond to the *instrument* that is being played. But now you input whatever MIDI notes have arrived at the current cursor position. Little bit confusingly, the player still respects the channel instrument as the instrument being played, so even if you input a note on track corresponding to instrument Foo, but the channel of the MIDI message corresponds to instrument Bar, the player will play instrument Bar. I guess at least we should at least somehow trigger the right instrument to not confuse the user?

But from your use case, I think you intentionally do not want to respect the channel information of the MIDI message, right? So you want to have a MIDI keyboard near by, and place the cursor where-ever, and input a note at that position, so it's almost similar as pressing a note on keyboard?

File: tracker/action.go
[the remark this answers is not in the document]
Comment (qm210) [reply tracker/action.go:1]: obvious mistake, sorry

File: tracker/gomidi/midi.go
[the remark this answers is not in the document]
Comment (qm210) [reply tracker/gomidi/midi.go:1]: obvious mistake, too

File: tracker/table.go (hunk 5 above)
Comment (qm210) [reply tracker/table.go]: no. I don't recall why I put that there

File: tracker/bool.go [hunk 6]
```diff
@@ -67,6 +69,7 @@ func (m *IsRecording) Bool() Bool  { return Bool{m} }
 func (m *IsRecording) Value() bool { return (*Model)(m).recording }
 func (m *IsRecording) setValue(val bool) {
 	m.recording = val
+	// this toggles the NoteEditor because it does not get updated real-time
```
[the remark this answers is not in the document]
Comment (qm210) [reply tracker/bool.go:72]: damn, I forgot this. it was somewhat related to the issue because at first, I did not understand why the panel was hidden while recording. I'll remove my comment.

File: tracker/table.go (hunk 3 above)
Comment (qm210) [reply tracker/table.go:152]: turns out that I accidentally triggered a GoLand recommendation. That's just polluting this pull request. Undid that.

File: cmd/sointu-track/main.go [hunk 7]
```diff
@@ -54,19 +45,23 @@ func main() {
 		recoveryFile = filepath.Join(configDir, "Sointu", "sointu-track-recovery")
 	}
 	midiContext := gomidi.NewContext()
+	defer midiContext.DestroyContext()
+	midiContext.TryOpenDefaultInputDevice(*defaultMidiInput, *firstMidiInput)
 	model, player := tracker.NewModelPlayer(cmd.MainSynther, midiContext, recoveryFile)
-	defer model.MIDI.Close()
+	fmt.Printf("Context created.\n")
```
[the remark this answers is not in the document]
Comment (qm210) [reply cmd/sointu-track/main.go]: check

File: tracker/action.go [hunk 8]
```diff
@@ -451,9 +463,8 @@ func (m *Model) completeAction(checkSave bool) {
 	}
 	switch m.dialog {
 	case NewSongChanges, NewSongSaveExplorer:
-		c := m.change("NewSong", SongChange, MajorChange)
+		c := m.change("NewSong", SongChange|LoopChange, MajorChange)
 		m.resetSong()
-		m.setLoop(Loop{})
```
[the remark this answers is not in the document]
Comment (qm210) [reply tracker/action.go]: mistake

File: go.sum [hunk 9]
```diff
@@ -21,6 +21,8 @@ github.com/davecgh/go-spew v1.1.0 h1:ZDRjVQ15GmhC3fiQ8ni8+OwkZQO4DARzQgrnXU1Liz8
 github.com/davecgh/go-spew v1.1.0/go.mod h1:J7Y8YcW2NihsgmVo/mv3lAwl/skON4iLHjSsI+c5H38=
 github.com/ebitengine/oto/v3 v3.3.0 h1:34lJpJLqda0Iee9g9p8RWtVVwBcOOO2YSIS2x4yD1OQ=
 github.com/ebitengine/oto/v3 v3.3.0/go.mod h1:MZeb/lwoC4DCOdiTIxYezrURTw7EvK/yF863+tmBI+U=
+github.com/ebitengine/oto/v3 v3.3.1 h1:d4McwGQuXOT0GL7bA5g9ZnaUEIEjQvG3hafzMy+T3qE=
```
[the remark this answers is not in the document]
Comment (qm210) [reply go.sum]: correct. mistake.

File: tracker/action.go [hunk 10]
```diff
@@ -433,14 +445,14 @@ func (m *Model) Cancel() Action      { return Allow(func() { m.dialog = NoDialog
...
 func (m *Model) ExportFloat() Action { return Allow(func() { m.dialog = ExportFloatExplorer }) }
 func (m *Model) ExportInt16() Action { return Allow(func() { m.dialog = ExportInt16Explorer }) }
-func (m *Model) SelectMidiInput(item MIDIDevice) Action {
+func (m *Model) SelectMidiInput(item MIDIDevicer) Action {
 	return Allow(func() {
-		if err := item.Open(); err != nil {
-			message := fmt.Sprintf("Could not open MIDI device: %s", item)
-			m.Alerts().Add(message, Error)
-		} else {
+		if m.MIDI.OpenInputDevice(item) {
```
[the remark this answers is not in the document]
Comment (qm210) [reply tracker/action.go]: mistake

File: song.go [hunk 11]
```diff
@@ -297,3 +297,30 @@ func (s *Song) Validate() error {
...
 	return nil
 }
+
+func (s *Song) InstrumentForTrack(trackIndex int) (int, bool) {
+	voiceIndex := s.Score.FirstVoiceForTrack(trackIndex)
+	instrument, err := s.Patch.InstrumentForVoice(voiceIndex)
+	return instrument, err == nil
+}
+
+func (s *Song) AllTracksWithSameInstrument(trackIndex int) []int {
```
[the remark this answers is not in the document]
Comment (qm210) [reply song.go]: I think I made it work :)

File: tracker/gioui/songpanel.go [hunk 12]
```diff
@@ -84,7 +84,7 @@ func NewSongPanel(model *tracker.Model) *SongPanel {
 		{IconBytes: icons.ContentRedo, Text: "Redo", ShortcutText: keyActionMap["Redo"], Doer: model.Redo()},
 		{IconBytes: icons.ImageCrop, Text: "Remove unused data", ShortcutText: keyActionMap["RemoveUnused"], Doer: model.RemoveUnused()},
 	}
-	for input := range model.MIDI.InputDevices {
```
[the remark this answers is not in the document]
Comment (qm210) [reply tracker/gioui/songpanel.go]: mistake

File: tracker/gioui/tracker.go [hunk 13]
```diff
@@ -306,3 +309,46 @@ func (t *Tracker) layoutTop(gtx layout.Context) layout.Dimensions {
...
+			return
+		}
+	}
+	t.MidiNotePlaying = append(t.MidiNotePlaying, note)
+}
+
+func (t *Tracker) removeFromMidiNotePlaying(note byte) {
+	for i, n := range t.MidiNotePlaying {
+		if n == note {
+			t.MidiNotePlaying = append(
```
[the remark this answers is not in the document]
Comment (qm210) [reply tracker/gioui/tracker.go:348]: no, I want to keep track of the order (e.g. if I press four keys, but have only three tracks available, I want to know which three were the first

File: tracker/model.go [hunk 14]
```diff
@@ -74,9 +75,10 @@ type (
 		synther sointu.Synther // the synther used to create new synths
 
 		PlayerMessages chan PlayerMsg
-		modelMessages  chan<- interface{}
+		ModelMessages  chan<- interface{}
 
-		MIDI MIDIContext
```
[the remark this answers is not in the document]
Comment (qm210) [reply tracker/model.go:79]: sure. mistake. sorry

File: tracker/gioui/tracker.go [hunk 15]
```diff
@@ -306,3 +309,46 @@ func (t *Tracker) layoutTop(gtx layout.Context) layout.Dimensions {
 		},
 	)
 }
+
+/// Event Handling (for UI updates when playing etc.)
+
+func (t *Tracker) ProcessMessage(msg interface{}) {
+	switch msg.(type) {
+	case tracker.StartPlayMsg:
+		fmt.Println("Tracker received StartPlayMsg")
```
[the remark this answers is not in the document]
Comment (qm210) [reply tracker/gioui/tracker.go:318]: kk

File: tracker/gioui/songpanel.go [hunk 16]
```diff
@@ -101,7 +101,6 @@ func NewSongPanel(model *tracker.Model) *SongPanel {
 	ret.followOffHint = makeHint("Follow off", " (%s)", "FollowToggle")
 	ret.loopOffHint = makeHint("Loop off", " (%s)", "LoopToggle")
 	ret.loopOnHint = makeHint("Loop on", " (%s)", "LoopToggle")
-
```
[the remark this answers is not in the document]
Comment (qm210) [reply tracker/gioui/songpanel.go]: yes

File: tracker/gioui/note_editor.go [hunk 17]
```diff
@@ -249,6 +261,12 @@ func (te *NoteEditor) layoutTracks(gtx C, t *Tracker) D {
 	drawSelection := te.scrollTable.Table.Cursor() != te.scrollTable.Table.Cursor2()
 	selection := te.scrollTable.Table.Range()
 
+	// TODO: maybe not best style to get the model value from the button, but it is bound to be the same...
+	hasTrackMidiIn := te.TrackMidiInBtn.Bool.Value()
+
+	// TODO @qm210: see table.go -> why defined for *Order, not *Model, when we need that cast anyway?
+	otherTracksForPolyphony := ((*tracker.Order)(t.Model)).TrackIndicesForCurrentInstrument()
+
```
[the remark this answers is not in the document]
Comment (qm210) [reply tracker/gioui/note_editor.go]: oh. didn't see that. it's still in there as t was (just moved, and without the comment), but I can try that in the future

**Comment by vsariola** [comment]
Merged, thank you for your contribution!
